astriaorg/astria
# feat: add an RFC 6962 compliant Merkle tree with flat memory representation

## Summary
Implement a RFC 6962 compliant merkle tree with flat memory representation, remove ct-merkle.

## Background
We want to use a merkle tree that minimizes allocations because constructing merkle trees is a hot operation that happens every time new blocks are generated or verified. Worse, for blocks containing transactions for several distinct rollups a merkle Tree is created per rollup. Finally, Astria wants a minimal API that is tailored exactly to its needs to make audits as simple as possible.

The implementation of this merkle tree library does not allocate space for its leaves but only stores leaf and branch hashes in a flat, one dimensional array. Furthermore, it only supports sha256 hashes. For background reading refer to the following links:

- https://datatracker.ietf.org/doc/html/rfc6962 (reference merkle tree this implementation adheres to)
- https://www.datprotocol.com/deps/0002-hypercore/
- https://datatracker.ietf.org/doc/rfc7574
- https://www.ietf.org/archive/id/draft-ietf-mls-protocol-14.html
- https://mmapped.blog/posts/22-flat-in-order-trees (this is reference this implementation builds on in terms of memory representation)

## Changes
- Add  an `astria-merkle` crate implementing a flat merkle tree.
- Rewrite all services/crates in terms of it.
- Simplify several utility functions by implementing APIs on the merkle tree, avoiding allocations.

## Testing
- Ensure the merkle tree passes all tests with known roots defined by https://github.com/transparency-dev/merkle
- Add unit and doc tests for all high level APIs.
- Ensure that all snapshot tests in astria sequencer still pass without changing the expdcted hashes

## Breaking Changelist
- Sequencer block data and sequencer/rollup blobs are changed, breaking compatibility with prior versions of sequencer, sequencer-relayer, conductor.
